offer-manager-app/src/main/groovy/life/qbic/portal/offermanager/offergeneration/OfferTemplate.groovy

```diff
@@ -0,0 +1,370 @@
+        }
+        document.select(".total-costs .overheads > .costs").each {element ->
+            element.text(Currency.format(offer.getOverheadTotal()))
+        }
+        document.select(".total-costs .vat > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalVat()))
+        }
+        document.select(".total-costs > .totals-section > .costs").each {element ->
+            element.text(Currency.format(offer.getTotalCost()))
+        }
+    }
+
+    private static void fillDataGenerationItems(Document document, OfferContent offer) {
+        Element dataGenerationTable = document.getElementById("data-generation-items")
+
+        // remove table if no items are available
+        if (offer.getDataGenerationItems().isEmpty()) {
+            dataGenerationTable.remove()
+            return
+        }
+        // clear the table
+        dataGenerationTable.select("tbody > tr").each {element ->
+            element.remove()
+        }
+        // fill the table
+        int position = 1
+        Element tableBody = dataGenerationTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataGenerationItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
+                tableBody.append(discountPosition.outerHtml())
+                position++
+            }
+        }
+        // set the footer
+        String netCosts = Currency.format(offer.getNetDataGeneration())
+        dataGenerationTable.select("> tfoot .costs").first().text(netCosts)
+    }
+    private static void fillDataAnalysisItems(Document document, OfferContent offer) {
+        Element dataAnalysisTable = document.getElementById("data-analysis-items")
+        // remove table if no items are available
+        if (offer.getDataAnalysisItems().isEmpty()) {
+            dataAnalysisTable.remove()
+            return
+        }
+        // clear the table
+        dataAnalysisTable.select("tbody > tr").each {element ->
+            element.remove()
+        }
+        // fill the table
+        int position = 1
+        Element tableBody = dataAnalysisTable.selectFirst("tbody")
+        for (OfferItem offerItem in offer.getDataAnalysisItems()) {
+            OfferPosition offerPosition = OfferPosition.createProductItem(position, offerItem)
+            tableBody.append(offerPosition.outerHtml())
+            position++
+            if (offerItem.getQuantityDiscount() > 0) {
+                OfferPosition discountPosition = OfferPosition.createDiscount(position, position - 1, offerItem)
+                tableBody.append(discountPosition.outerHtml())
+                position++
+            }
+        }
```

Comment:
This is repeated for each table. I think you could outsource that to a method and call it for each of the product groups

Reply:
You have got a point here :+1: I will do that.